engine/execution/storehouse/checkpoint_validator.go

```diff
@@ -0,0 +1,137 @@
+package storehouse
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/rs/zerolog"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/onflow/flow-go/engine/execution"
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/common/convert"
+	"github.com/onflow/flow-go/ledger/complete/wal"
+	"github.com/onflow/flow-go/storage"
+)
+
+// ValidateWithCheckpoint validates the registers in the given store against the leaf nodes read from the checkpoint file.
+// Limitation: the validation can not cover if there are extra non-empty registers in the store that are not in the checkpoint file.
+func ValidateWithCheckpoint(
+	log zerolog.Logger,
+	ctx context.Context,
+	store execution.OnDiskRegisterStore,
+	results storage.ExecutionResults,
+	headers storage.Headers,
+	checkpointDir string, // checkpointDir must have a root.checkpoint file that contains only a single trie
+	blockHeight uint64,
+	workerCount int,
+) error {
+	// used by the wal reader to send leaf nodes read from checkpoint file
+	// used by N workers to validate registers in store
+	leafNodeChan := make(chan *wal.LeafNode, 1000)
+
+	// create N workers to validate registers in store
+	cct, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	g, gCtx := errgroup.WithContext(cct)
+
+	start := time.Now()
+	log.Info().Msgf("validation registers from checkpoint with %v worker", workerCount)
+	for i := 0; i < workerCount; i++ {
+		g.Go(func() error {
+			return validatingRegisterInStore(gCtx, store, leafNodeChan, blockHeight)
+		})
+	}
+
+	rootHash, err := rootHashByHeight(results, headers, blockHeight)
+	if err != nil {
+		return err
+	}
+
+	// read leaf nodes from checkpoint file and send to leafNodeChan
+	err = wal.OpenAndReadLeafNodesFromCheckpointV6(leafNodeChan, checkpointDir, "root.checkpoint", rootHash, log)
+	if err != nil {
+		return fmt.Errorf("error reading leaf node from checkpoint: %w", err)
+	}
```

Comment:
I think it is kind of a big limitation to only support validating payloads against "root.checkpoint". 

Maybe we can support validation against checkpoint files and wal files by:
- Create multiple tries by loading latest checkpoint files and wal files
- Get a single trie by specified root hash
- Get all payloads from the single trie
- Validate all payloads against storehouse

This approach would remove the "root.checkpoint" limitation, and offer more flexibility to validate against any stored tries.

Thoughts?